diff --git a/osu.Game.Rulesets.Osu/Difficulty/Evaluators/Speed/SpeedEvaluator.cs b/osu.Game.Rulesets.Osu/Difficulty/Evaluators/Speed/SpeedEvaluator.cs
index 3e3b63e924..34191d27c3 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Evaluators/Speed/SpeedEvaluator.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Evaluators/Speed/SpeedEvaluator.cs
@@ -14,7 +14,8 @@ public static class SpeedEvaluator
     {
         private const double min_speed_bonus = 200; // 200 BPM 1/4th
         private const double speed_balancing_factor = 40;
-
+        private const double single_spacing_threshold = OsuDifficultyHitObject.NORMALISED_DIAMETER * 1.25; // 1.25 circles distance between centers
+        private const double distance_multiplier = 0.8;
         /// <summary>
         /// Evaluates the difficulty of tapping the current object, based on:
         /// <list type="bullet">
@@ -22,12 +23,13 @@ public static class SpeedEvaluator
         /// <item><description>and how easily they can be cheesed.</description></item>
         /// </list>
         /// </summary>
-        public static double EvaluateDifficultyOf(DifficultyHitObject current)
+        public static double EvaluateDifficultyOf(DifficultyHitObject current, bool WithDistance)
         {
             if (current.BaseObject is Spinner)
                 return 0;
 
             var osuCurrObj = (OsuDifficultyHitObject)current;
+            var osuPrevObj = current.Index > 0 ? (OsuDifficultyHitObject)current.Previous(0) : null;
 
             double strainTime = osuCurrObj.AdjustedDeltaTime;
             double doubletapness = 1.0 - osuCurrObj.GetDoubletapness((OsuDifficultyHitObject?)osuCurrObj.Next(0));
@@ -43,8 +45,23 @@ public static double EvaluateDifficultyOf(DifficultyHitObject current)
             if (DifficultyCalculationUtils.MillisecondsToBPM(strainTime) > min_speed_bonus)
                 speedBonus = 0.75 * Math.Pow((DifficultyCalculationUtils.BPMToMilliseconds(min_speed_bonus) - strainTime) / speed_balancing_factor, 2);
 
+            double travelDistance = osuPrevObj?.TravelDistance ?? 0;
+            double distance = travelDistance + osuCurrObj.MinimumJumpDistance;
+
+            // Cap distance at single_spacing_threshold
+            distance = Math.Min(distance, single_spacing_threshold);
+
+            // Max distance bonus is 1 * `distance_multiplier` at single_spacing_threshold
+            double distanceBonus = Math.Pow(distance / single_spacing_threshold, 3.95) * distance_multiplier;
+
+            // Apply reduced small circle bonus because flow aim difficulty on small circles doesn't scale as hard as jumps
+            distanceBonus *= Math.Sqrt(osuCurrObj.SmallCircleBonus);
+
+            if (!WithDistance)
+                distanceBonus = 0;
+
             // Base difficulty with all bonuses
-            double speedDifficulty = (1 + speedBonus) * 1000 / strainTime;
+            double speedDifficulty = (1 + speedBonus + distanceBonus) * 1000 / strainTime;
 
             speedDifficulty *= highBpmBonus(osuCurrObj.AdjustedDeltaTime);
 
diff --git a/osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs b/osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs
index eed6a87d84..0743959229 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/OsuDifficultyCalculator.cs
@@ -35,7 +35,7 @@ protected override DifficultyAttributes CreateDifficultyAttributes(IBeatmap beat
 
             var aim = skills.OfType<Aim>().Single(a => a.IncludeSliders);
             var aimWithoutSliders = skills.OfType<Aim>().Single(a => !a.IncludeSliders);
-            var speed = skills.OfType<Speed>().Single();
+            var speed = skills.OfType<Speed>().First();
             var flashlight = skills.OfType<Flashlight>().SingleOrDefault();
             var reading = skills.OfType<Reading>().Single();
 
@@ -166,7 +166,8 @@ protected override Skill[] CreateSkills(IBeatmap beatmap, Mod[] mods)
             {
                 new Aim(mods, true),
                 new Aim(mods, false),
-                new Speed(mods),
+                new Speed(mods, true),
+                new Speed(mods, false),
                 new Reading(mods)
             };
 
diff --git a/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs b/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
index bc0383f074..860c4efd60 100644
--- a/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
+++ b/osu.Game.Rulesets.Osu/Difficulty/Skills/Speed.cs
@@ -20,6 +20,7 @@ namespace osu.Game.Rulesets.Osu.Difficulty.Skills
     /// </summary>
     public class Speed : HarmonicSkill
     {
+        public bool WithDistance { get; }
         private double skillMultiplier => 1.16;
 
         private readonly List<double> sliderStrains = new List<double>();
@@ -31,9 +32,10 @@ public class Speed : HarmonicSkill
         protected override double HarmonicScale => 20;
         protected override double DecayExponent => 0.9;
 
-        public Speed(Mod[] mods)
+        public Speed(Mod[] mods, bool withDistance)
             : base(mods)
         {
+            WithDistance = withDistance;
         }
 
         private double strainDecay(double ms) => Math.Pow(strainDecayBase, ms / 1000);
@@ -60,7 +62,7 @@ protected override double ObjectDifficultyOf(DifficultyHitObject current)
 
         private double calculateAdjustedDifficulty(DifficultyHitObject current)
         {
-            double difficulty = SpeedEvaluator.EvaluateDifficultyOf(current);
+            double difficulty = SpeedEvaluator.EvaluateDifficultyOf(current, WithDistance);
 
             if (Mods.Any(m => m is OsuModAutopilot))
                 difficulty *= 0.5;
